<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Functional enrichment analysis of high-throughput omics data • enrichOmics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Functional enrichment analysis of high-throughput omics data">
<meta property="og:description" content="Overview of existing methods for enrichment analysis of gene
    expression data with regard to functional gene sets, pathways, and networks.
    Functionality for differential expression analysis, set- and network-based
    enrichment analysis, along with visualization and exploration of results.
    Introduction to concepts for enrichment analysis of genomic regions and
    regulatory elements. Outlook on current developments to extend gene set
    enrichment analysis for data derived from multiple high-throughput omics
    assays. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">enrichOmics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/Geistlinger_enrichOmics.html">Workshop</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/waldronlab/enrichOmics">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="functional-enrichment-analysis-of-high-throughput-omics-data" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#functional-enrichment-analysis-of-high-throughput-omics-data" class="anchor"></a>Functional enrichment analysis of high-throughput omics data</h1></div>
</div>
<div id="instructors-names-and-contact-information" class="section level1">
<h1 class="hasAnchor">
<a href="#instructors-names-and-contact-information" class="anchor"></a>Instructor(s) name(s) and contact information</h1>
<p>Ludwig Geistlinger and Levi Waldron</p>
<p>CUNY School of Public Health 55 W 125th St, New York, NY 10027</p>
<p><a href="mailto:Ludwig.Geistlinger@sph.cuny.edu" class="email">Ludwig.Geistlinger@sph.cuny.edu</a></p>
</div>
<div id="workshop-description" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h1>
<p>This workshop gives an in-depth overview of existing methods for enrichment analysis of gene expression data with regard to functional gene sets, pathways, and networks. The workshop will help participants understand the distinctions between assumptions and hypotheses of existing methods as well as the differences in objectives and interpretation of results. It will provide code and hands-on practice of all necessary steps for differential expression analysis, gene set- and network-based enrichment analysis, and identification of enriched genomic regions and regulatory elements, along with visualization and exploration of results. A previous well-attended (&gt;50 participants) version of this workshop was presented at Bioc2017 (<a href="https://www.bioconductor.org/help/course-materials/2017/BioC2017/Day1/Workshops/OmicsData/doc/enrichOmics.html" class="uri">https://www.bioconductor.org/help/course-materials/2017/BioC2017/Day1/Workshops/OmicsData/doc/enrichOmics.html</a>).</p>
<div id="pre-requisites" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h2>
<ul>
<li><p>Basic knowledge of R syntax</p></li>
<li><p>Familiarity with the SummarizedExperiment class</p></li>
<li><p>Familiarity with the GenomicRanges class</p></li>
<li><p>Familiarity with high-throughput gene expression data as obtained with microarrays and RNA-seq</p></li>
<li><p>Familiarity with the concept of differential expression analysis (with e.g. limma, edgeR, DESeq2)</p></li>
</ul>
</div>
<div id="workshop-participation" class="section level2">
<h2 class="hasAnchor">
<a href="#workshop-participation" class="anchor"></a>Workshop Participation</h2>
<p>Execution of example code and hands-on practice</p>
</div>
<div id="r--bioconductor-packages-used" class="section level2">
<h2 class="hasAnchor">
<a href="#r--bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h2>
<ul>
<li>EnrichmentBrowser</li>
<li>regioneR</li>
</ul>
</div>
<div id="time-outline" class="section level2">
<h2 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h2>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Background</td>
<td>30m</td>
</tr>
<tr class="even">
<td>Differential expression analysis</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Gene set analysis</td>
<td>30m</td>
</tr>
<tr class="even">
<td>Gene network analysis</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Genomic region analysis</td>
<td>15m</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="goals-and-objectives" class="section level1">
<h1 class="hasAnchor">
<a href="#goals-and-objectives" class="anchor"></a>Goals and objectives</h1>
<p>Theory</p>
<ul>
<li>Gene sets, pathways &amp; regulatory networks</li>
<li>Resources</li>
<li>Gene set analysis vs. gene set enrichment analysis</li>
<li>Underlying null: competitive vs. self-contained</li>
<li>Generations: ora, fcs &amp; topology-based</li>
</ul>
<p>Practice:</p>
<ul>
<li>Data types: microarray vs. RNA-seq</li>
<li>Differential expression analysis</li>
<li>Defining gene sets according to GO and KEGG</li>
<li>GO/KEGG overrepresentation analysis</li>
<li>Functional class scoring &amp; permutation testing</li>
<li>Network-based enrichment analysis</li>
<li>Genomic region enrichment analysis</li>
</ul>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/Artistic-2.0">Artistic-2.0</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Ludwig Geistlinger <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Ludwig Geistlinger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
